File: experiments/g1/schema.py
"""Schema hash and domain-separation tags (canonical-data-v1).

The SchemaHash is the SHA-256 of the *raw bytes* of the frozen schema file —
no language may reformat the JSON before hashing. It is split into two 128-bit
halves so it fits cleanly into BN254 field elements. Domain tags are
``SHA-256(name) mod p``.
"""
from __future__ import annotations

import hashlib
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[2]
SCHEMA_PATH = REPO_ROOT / "specs" / "canonical-data-v1.schema.json"

DOMAIN_NAMES = (
    "DDTM_ROW_V1",
    "DDTM_PADDING_V1",
    "DDTM_NODE_V1",
    "DDTM_SCHEMA_V1",
)


def schema_digest(path: Path = SCHEMA_PATH) -> bytes:
    """Raw 32-byte SHA-256 of the canonical schema file bytes."""
    return hashlib.sha256(path.read_bytes()).digest()


def schema_halves(path: Path = SCHEMA_PATH) -> tuple[int, int]:
    """Return ``(schemaHi, schemaLo)`` as two 128-bit big-endian integers."""
    digest = schema_digest(path)
    schema_hi = int.from_bytes(digest[:16], "big")
    schema_lo = int.from_bytes(digest[16:], "big")
    return schema_hi, schema_lo


def schema_sha256_hex(path: Path = SCHEMA_PATH) -> str:
    return schema_digest(path).hex()


def domain_tags(modulus: int) -> dict[str, int]:
    """Compute every domain tag as ``SHA-256(name) mod modulus``."""
    return {
        name: int.from_bytes(hashlib.sha256(name.encode("ascii")).digest(), "big")
        % modulus
        for name in DOMAIN_NAMES
    }

File: canonicalizer-go/internal/canonical/schema.go
package canonical

import (
	"crypto/sha256"
	"encoding/hex"
	"math/big"
	"os"

	"github.com/consensys/gnark-crypto/ecc/bn254/fr"
)

// 域标签来源名（domain-tags-v1）。
var domainNames = []string{
	"DDTM_ROW_V1",
	"DDTM_PADDING_V1",
	"DDTM_NODE_V1",
	"DDTM_SCHEMA_V1",
}

// SchemaDigest 返回规范化 schema 文件原始字节的 SHA-256。
func SchemaDigest(path string) ([]byte, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	sum := sha256.Sum256(raw)
	return sum[:], nil
}

// SchemaHalves 返回 (hi, lo)——两个 128 位大端域整数。
func SchemaHalves(path string) (fr.Element, fr.Element, error) {
	digest, err := SchemaDigest(path)
	if err != nil {
		return fr.Element{}, fr.Element{}, err
	}
	var hi, lo fr.Element
	hi.SetBigInt(new(big.Int).SetBytes(digest[:16]))
	lo.SetBigInt(new(big.Int).SetBytes(digest[16:]))
	return hi, lo, nil
}

// SchemaSHA256Hex 返回 schema 摘要的十六进制编码。
func SchemaSHA256Hex(path string) (string, error) {
	d, err := SchemaDigest(path)
	if err != nil {
		return "", err
	}
	return hex.EncodeToString(d), nil
}

// DomainTags 返回每个标签名的 SHA-256(name) mod p。
func DomainTags(modulus *big.Int) map[string]fr.Element {
	_ = modulus // 模数已由 fr.Element 隐含
	tags := make(map[string]fr.Element, len(domainNames))
	for _, name := range domainNames {
		sum := sha256.Sum256([]byte(name))
		var e fr.Element
		e.SetBigInt(new(big.Int).SetBytes(sum[:]))
		tags[name] = e
	}
	return tags
}
